File: apps/ios/Packages/DropCore/Sources/DropCore/SupabaseNotesRepository.swift
import Foundation
import Supabase

/// `notes_repository.dart`의 Supabase 구현을 옮긴 것.
public struct SupabaseNotesRepository: NotesRepository {
    private let client: SupabaseClient

    public init(client: SupabaseClient) {
        self.client = client
    }

    public func loadNotes() async throws -> [Note] {
        let notes: [Note] = try await run {
            try await client.from("notes").select().order("created_at", ascending: false).execute().value
        }
        guard !notes.isEmpty else { return [] }

        let ids = notes.map(\.id)
        let attachments: [Attachment] = try await run {
            try await client.from("attachments")
                .select()
                .in("note_id", values: ids)
                .order("created_at")
                .execute().value
        }
        let tagsByNoteID = try await loadTagsByNoteID(noteIDs: ids)

        return NoteAssembler.sorted(
            NoteAssembler.assemble(notes: notes, attachments: attachments, tagsByNoteID: tagsByNoteID)
        )
    }

    public func createNote(content: String, parentID: String?) async throws -> Note {
        // user_id는 클라이언트가 넣지 않는다 — RLS 정책과 DB 기본값이 정한다.
        guard client.auth.currentUser != nil else { throw NotesRepositoryError.notAuthenticated }

        return try await run {
            try await client.from("notes")
                .insert(NoteInsert(content: content, parentId: parentID, source: "mobile"))
                .select()
                .single()
                .execute().value
        }
    }

    public func updateNote(id: String, content: String) async throws {
        try await update(id: id, values: ["content": .string(content)])
    }

    public func moveToTrash(id: String) async throws {
        // 보관 해제를 함께 하지 않으면 휴지통과 보관함 양쪽에 나타난다.
        try await update(id: id, values: [
            "is_deleted": .bool(true),
            "deleted_at": .string(nowTimestamp()),
            "archived_at": .null,
        ])
    }

    public func restoreFromTrash(id: String) async throws {
        try await update(id: id, values: ["is_deleted": .bool(false), "deleted_at": .null])
    }

    public func archive(id: String) async throws {
        try await update(id: id, values: ["archived_at": .string(nowTimestamp())])
    }

    public func unarchive(id: String) async throws {
        try await update(id: id, values: ["archived_at": .null])
    }

    public func deletePermanently(id: String) async throws {
        _ = try await run { try await client.from("notes").delete().eq("id", value: id).execute() }
    }

    public func emptyTrash() async throws {
        guard let user = client.auth.currentUser else { throw NotesRepositoryError.notAuthenticated }
        _ = try await run {
            try await client.from("notes")
                .delete()
                .eq("user_id", value: user.id.uuidString)
                .not("deleted_at", operator: .is, value: AnyJSON.null)
                .execute()
        }
    }

    public func setPinned(id: String, isPinned: Bool) async throws {
        try await update(id: id, values: [
            "is_pinned": .bool(isPinned),
            "pinned_at": isPinned ? .string(nowTimestamp()) : .null,
        ])
    }

    public func setLocked(id: String, isLocked: Bool) async throws {
        try await update(id: id, values: ["is_locked": .bool(isLocked)])
    }

    public func setPriority(id: String, priority: Int) async throws {
        // DB CHECK 제약(0-3)에 걸리지 않도록 클라이언트에서 먼저 자른다.
        try await update(id: id, values: ["priority": .integer(min(max(priority, 0), 3))])
    }

    public func updateCategories(id: String, hasLink: Bool, hasMedia: Bool, hasFiles: Bool) async throws {
        try await update(id: id, values: [
            "has_link": .bool(hasLink),
            "has_media": .bool(hasMedia),
            "has_files": .bool(hasFiles),
        ])
    }

    // MARK: - 내부

    private func loadTagsByNoteID(noteIDs: [String]) async throws -> [String: [Tag]] {
        let rows: [NoteTagRow] = try await run {
            try await client.from("note_tags")
                .select("note_id, tags(*)")
                .in("note_id", values: noteIDs)
                .execute().value
        }
        return rows.reduce(into: [:]) { result, row in
            result[row.noteID, default: []].append(row.tag)
        }
    }

    private func update(id: String, values: [String: AnyJSON]) async throws {
        _ = try await run { try await client.from("notes").update(values).eq("id", value: id).execute() }
    }

    private func nowTimestamp() -> String {
        ISO8601DateFormatter().string(from: Date())
    }

    /// Supabase/PostgREST 오류를 화면이 다룰 수 있는 형태로 좁힌다.
    private func run<T>(_ operation: () async throws -> T) async throws -> T {
        do {
            return try await operation()
        } catch let error as PostgrestError {
            throw NotesRepositoryError.rejected(error.message)
        } catch let error as DecodingError {
            throw NotesRepositoryError.decoding(String(describing: error))
        } catch {
            throw NotesRepositoryError.network(error.localizedDescription)
        }
    }
}

private struct NoteInsert: Encodable {
    let content: String
    let parentId: String?
    let source: String
}

private struct NoteTagRow: Decodable {
    let noteID: String
    let tag: Tag

    private enum CodingKeys: String, CodingKey {
        case noteID = "noteId"
        case tag = "tags"
    }
}

File: apps/mobile/packages/drop_core/lib/src/supabase/supabase_notes_repository.dart
import '../attachment.dart';
import '../drop_json.dart';
import '../note.dart';
import '../note_assembler.dart';
import '../notes_repository.dart';
import '../repository_error.dart';
import '../tag.dart';
import 'insert_payloads.dart';
import 'supabase_rest.dart';

/// `notes` 테이블에 붙는 구현. DropCore `SupabaseNotesRepository.swift` 대응.
class SupabaseNotesRepository implements NotesRepository {
  final SupabaseRestClient client;

  SupabaseNotesRepository({required this.client});

  @override
  Future<List<Note>> loadNotes() async {
    final rows = await client.select('notes', {
      'select': '*',
      'order': 'created_at.desc',
    });
    final notes = await client.run(() async =>
        rows.map((row) => Note.fromJson((row as Map).cast())).toList());
    // 노트가 없으면 첨부·태그 쿼리를 아예 보내지 않는다.
    // 빈 `in` 필터는 PostgREST에서 오류이거나 전체 조회가 되어버린다.
    if (notes.isEmpty) return [];

    final ids = notes.map((note) => note.id).toList();
    final attachmentRows = await client.select('attachments', {
      'select': '*',
      'note_id': SupabaseRestClient.inFilter(ids),
      'order': 'created_at.asc',
    });
    final attachments = await client.run(() async => attachmentRows
        .map((row) => Attachment.fromJson((row as Map).cast()))
        .toList());
    final tagsByNoteId = await _loadTagsByNoteId(ids);

    return NoteAssembler.sorted(NoteAssembler.assemble(
      notes: notes,
      attachments: attachments,
      tagsByNoteId: tagsByNoteId,
    ));
  }

  @override
  Future<Note> createNote({String content = '', String? parentId}) async {
    // user_id를 반드시 실어 보낸다. INSERT 정책이 user_id = auth.uid()를 요구하는데
    // 컬럼에 기본값이 없어서, 빠뜨리면 NULL이 들어가 RLS가 거부한다.
    final userId = client.session.currentUserId;
    if (userId == null) throw const NotesRepositoryError.notAuthenticated();

    final row = await client.insertReturning(
      'notes',
      noteInsertPayload(content: content, parentId: parentId, userId: userId),
    );
    return client.run(() async => Note.fromJson(row));
  }

  @override
  Future<void> updateNote({required String id, required String content}) =>
      _update(id, {'content': content});

  @override
  Future<void> moveToTrash(String id) =>
      // Rule B (BRU-115): 복원은 받은편지함으로. 보관을 남기면 휴지통·보관함 양쪽에 나타난다.
      _update(id, {
        'is_deleted': true,
        'deleted_at': _now(),
        'archived_at': null,
      });

  @override
  Future<void> restoreFromTrash(String id) => _update(id, {
        'is_deleted': false,
        'deleted_at': null,
        'archived_at': null,
      });

  @override
  Future<void> archive(String id) => _update(id, {'archived_at': _now()});

  @override
  Future<void> unarchive(String id) => _update(id, {'archived_at': null});

  @override
  Future<void> deletePermanently(String id) =>
      client.delete('notes', filters: {'id': 'eq.$id'});

  @override
  Future<void> emptyTrash() async {
    final userId = client.session.currentUserId;
    if (userId == null) throw const NotesRepositoryError.notAuthenticated();
    await client.delete('notes', filters: {
      'user_id': 'eq.$userId',
      'deleted_at': 'not.is.null',
    });
  }

  @override
  Future<void> setPinned(String id, {required bool isPinned}) => _update(id, {
        'is_pinned': isPinned,
        'pinned_at': isPinned ? _now() : null,
      });

  @override
  Future<void> setLocked(String id, {required bool isLocked}) =>
      _update(id, {'is_locked': isLocked});

  @override
  Future<void> setPriority(String id, int priority) =>
      // DB CHECK 제약(0-3)에 걸리지 않도록 클라이언트에서 먼저 자른다.
      _update(id, {'priority': priority.clamp(0, 3)});

  @override
  Future<void> setType(String id, NoteType type) => _update(id, {
        'type': type.name,
        // 일반 노트로 되돌릴 때 완료 시각을 같은 갱신에서 지운다. 따로 두 번
        // 쓰면 그 사이 상태가 CHECK를 위반해 첫 갱신부터 거부된다 (BRU-184).
        if (type != NoteType.todo) 'completed_at': null,
      });

  @override
  Future<void> setCompleted(String id, {required bool completed}) => _update(
        id,
        {'completed_at': completed ? DateTime.now().toUtc().toIso8601String() : null},
      );

  @override
  Future<void> updateCategories(
    String id, {
    required bool hasLink,
    required bool hasMedia,
    required bool hasFiles,
  }) =>
      _update(id, {
        'has_link': hasLink,
        'has_media': hasMedia,
        'has_files': hasFiles,
      });

  // 내부

  Future<Map<String, List<Tag>>> _loadTagsByNoteId(List<String> noteIds) async {
    final rows = await client.select('note_tags', {
      'select': 'note_id, tags(*)',
      'note_id': SupabaseRestClient.inFilter(noteIds),
    });
    return client.run(() async {
      final result = <String, List<Tag>>{};
      for (final row in rows) {
        final map = (row as Map).cast<String, Object?>();
        final noteId = map['note_id'] as String;
        final tag = Tag.fromJson((map['tags'] as Map).cast());
        result.putIfAbsent(noteId, () => []).add(tag);
      }
      return result;
    });
  }

  Future<void> _update(String id, Map<String, Object?> values) =>
      client.update('notes', values, filters: {'id': 'eq.$id'});

  static String _now() => formatPostgresTimestamp(DateTime.now());
}
